src/edu/washington/escience/myria/operator/SampleWoR.java

```diff
@@ -0,0 +1,66 @@
+package edu.washington.escience.myria.operator;
+
+import java.util.BitSet;
+import java.util.Random;
+
+import com.google.common.base.Preconditions;
+import edu.washington.escience.myria.storage.TupleBatch;
+
+public class SampleWoR extends Sample {
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /** True if operator has extracted sampling info. */
+  private boolean computedSamplingInfo = false;
+
+  /** Current number of samples that have been selected. */
+  private int samplesAcquired = 0;
+
+  /** Current number of tuples that have been processed. */
+  private int tuplesProcessed = 0;
+
+  public SampleWoR(final Operator left, final Operator right) {
+    super(left, right);
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws Exception {
+    // Extract sampling info from left operator.
+    if (!computedSamplingInfo) {
+      TupleBatch tb = getLeft().nextReady();
+      if (tb == null)
+        return null;
+      extractSamplingInfo(tb);
+
+      // Cannot sampleWoR more tuples than there are.
+      Preconditions.checkState(sampleSize <= populationSize,
+          "Cannot SampleWoR %s tuples from a population of size %s",
+          sampleSize, populationSize);
+
+      getLeft().close();
+      computedSamplingInfo = true;
+    }
+
+    // SampleWoR from the right operator.
```

Comment:
This algorithm is standard, but is linear in the population size rather than the sample size. One algorithm for sampling WoR that is linear in the sample size is Floyd's algorithm from _Programming Pearls_ (`N` is population size, `M` is sample size):
```
initialize set S to empty
for J := N-M + 1 to N do
  T := RandInt(1, J)
  if T is not in S then
    insert T in S
  else
    insert J in S
```
This is tricky but can be proven with a simple induction (which just took me an embarrassingly long time to prove).

Anyway, not sure if this is a valid performance concern at all, but just thought I'd mention it. I also wonder how this works for populations whose size isn't known (e.g., streaming intermediate results). Is that a valid scenario? If so, I guess some variant of reservoir sampling would be called for...